crates/environ/src/component/translate.rs

```diff
@@ -219,6 +225,44 @@ struct LocalCanonicalOptions {
     post_return: Option<FuncIndex>,
 }
 
+#[derive(Copy, Clone)]
+enum ComponentType {
+    /// A component was imported from so its type is listed by an index.
+    Index(TypeComponentIndex),
+    /// A component was defined statically inline so its type information can be
+    /// found in the `Translation` itself.
+    Static(StaticComponentIndex),
+}
+
+#[derive(Copy, Clone)]
+enum ComponentInstanceType {
+    /// An instance was imported so its type information is listed by index.
+    Index(TypeComponentInstanceIndex),
+    /// An instance was created through instantiating an imported component, so
+    /// the type of the index is inferred from the type of the component.
+    InstantiatedIndex(TypeComponentIndex),
+    /// An instance was created through instantiation of a statically defined
+    /// component, so the type information is inferred from a `Translation`.
+    InstantiatedStatic(StaticComponentIndex),
+    /// An instance was created from a list of other items, and the list can be
+    /// found in the `synthetic_instance_types` map.
+    Synthetic(SyntheticInstanceTypeIndex),
+}
+
+/// Only used as part of `synthetic_instance_types` which is used to infer type
+/// information for component-model items. This intentionally doesn't cover
+/// everything (e.g. modules are missing at this time since they aren't needed).
+#[derive(Copy, Clone)]
+enum ComponentItemType {
+    Func(TypeFuncIndex),
+    Component(ComponentType),
+    Instance(ComponentInstanceType),
+}
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+struct SyntheticInstanceTypeIndex(u32);
+cranelift_entity::entity_impl!(SyntheticInstanceTypeIndex);
```

Comment:
This is a private indexing type where I basically just didn't want to use `usize` and isn't otherwise used outside of this file, so I didn't want to put it in that public location.